File: back/src/main/java/com/openclassrooms/mddapi/services/impl/ArticleService.java
package com.openclassrooms.mddapi.services.impl;

import com.openclassrooms.mddapi.exceptions.ApiException;
import com.openclassrooms.mddapi.models.Article;
import com.openclassrooms.mddapi.models.Comment;
import com.openclassrooms.mddapi.models.Topic;
import com.openclassrooms.mddapi.repositories.IArticleRepository;
import com.openclassrooms.mddapi.repositories.ICommentRepository;
import com.openclassrooms.mddapi.services.IArticleService;
import com.openclassrooms.mddapi.services.ITopicService;
import com.openclassrooms.mddapi.services.IUserService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.util.List;

@RequiredArgsConstructor
@Service
public class ArticleService implements IArticleService {

    private final ITopicService topicService;
    private final IUserService userService;
    private final IArticleRepository articleRepository;
    private final ICommentRepository commentRepository;

    @Override
    public Article createArticle(String title, String content, long topicId) {
        Topic topic = topicService.findById(topicId);

        Article article = Article.builder()
                .topic(topic)
                .title(title)
                .content(content)
                .author(userService.me())
                .comments(List.of())
                .build();

        return articleRepository.save(article);
    }

    @Override
    public List<Article> getArticles() {
        return articleRepository.findAll();
    }

    @Override
    public Article getArticle(long id) {
        return articleRepository.findById(id).orElseThrow(() -> new ApiException(HttpStatus.NOT_FOUND, "Article non trouvé."));
    }

    @Override
    public Article createComment(long id, String content) {
        Comment comment = Comment.builder()
                .content(content)
                .author(userService.me())
                .article(getArticle(id))
                .build();

        commentRepository.save(comment);

        return getArticle(id);
    }
}

File: front/src/app/http/article.service.ts
import {Injectable} from '@angular/core';
import {ApiService} from './api.service';
import {Observable} from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class ArticleService {

  constructor(private apiService: ApiService) {
  }

  /**
   * Create an article
   * @param topicId The topic id
   * @param title The article title
   * @param content The article content
   */
  createArticle(topicId: number, title: string, content: string): Observable<void> {
    return this.apiService.post('/articles', {
      topicId,
      title,
      content,
    });
  }

  /**
   * Get the articles from the server
   */
  getArticles(): Observable<Article[]> {
    return this.apiService.get<Article[]>('/articles');
  }

  /**
   * Get an article by id
   * @param id The article id
   */
  getArticle(id: number): Observable<Article> {
    return this.apiService.get<Article>('/articles/' + id);
  }

  /**
   * Like an article
   * @param articleId The article id
   * @param content The article content
   */
  createComment(articleId: number, content: string): Observable<void> {
    return this.apiService.post('/articles/' + articleId + '/comments', {
      content,
    });
  }

}
